### ISSUE TITLE
go.dev: support packages with unknown/unsupported licenses

### ISSUE THREAD

**Issue by eliasnaur** [golang/go#36840]
There has been work on this, but @mvdan suggested I create an issue for it for tracking.

The motivating example is the dual-licensed (MIT/UNLICENSE) gioui.org packages. See https://pkg.go.dev/gioui.org/op. The problem is that go.dev refuses to display certain meta-data if the license is not detected or in a list of supported licenses. In particular, the blocking of "Doc" and "Overview" sections make the pkg.go.dev listing almost useless.

Compare godoc.org which, while inferior to pkg.go.dev in other ways, has no problem displaying documentation regardless of licenses.

The Go team seems to have a manual process of vetting packages such as gioui.org. However, that process is slow for all parties and might not cover every weird license out there.

@rogpeppe because pkg.go.dev also blocks some of his packages.

**Comment by julieqiu** [comment]
We discussed this on the golang-tools call - but posting here for posterity.

At the moment, we can only post metadata about a package/module (imports, importers, published date, etc.) when it does not have a recognized license per the requirements at https://pkg.go.dev/license-policy.

**Comment by mvdan** [comment]
I think Julie misclicked - from her comment, I don't understand why we should close this issue. It seems useful to leave it open, so that other module authors having similar issues can subscribe for updates.

**Comment by julieqiu** [comment]
I closed this issue because we don't intend to support unknown/unsupported licenses in the foreseeable future.

The most up-to-date information on our license policy will be at https://pkg.go.dev/license-policy. Authors can feel free to file an issue, or email go-discovery-feedback@google.com for any questions about their specific package.

Any major updates will be posted on the [tracking issue for pkg.go.dev](https://golang.org/s/discovery-updates), so folks should subscribe to that issue for project updates.

**Comment by eliasnaur** [comment]
Would you consider alternatives? It seems to me the issue that go.dev needs permission to a very limited use of the copyrighted work, namely displaying the textual documentation embedded in the repository. For example, how about a file that clearly states go.dev (or source code explorers in general) permission to display the package contents? Say, `<reporoot>/.allow-sourceexplorers`.

**Comment by eliasnaur** [comment]
FWIW, other source explorers such as https://docs.rs/ doesn't seem to be affected by licensing issues either.

**Comment by julieqiu** [comment]
We (and the proxy team) considered that option and decided that it wasn't viable. We would rather not solve this problem with a "just for us" solution, because it doesn't encourage package authors to fix their licenses for the broader open source ecosystem if there are issues, and other users have concerns about a package or module's licenses.

/cc @jba @findleyr @katiehockman @heschik @hana

**Comment by eliasnaur** [comment]
> a "just for us" solution

I deliberately named the file without reference to go.dev. I assumed that if Google has this legal issue with displaying source code meta-data, other similar explorers have it as well. Otherwise, what's special about Google?

> package authors to fix their licenses for the broader open source ecosystem if there are issues, and other users have concerns about a package or module's licenses.

I'm sorry, but I can't see what I can do to "fix my license" for the gioui.org modules. If anything, go.dev is being overly restrictive in not accepting dual licenses as OR, rather than AND (gioui.org is dual licensed MIT OR UNLICENSE and MIT is supported by go.dev).

Secondly, I don't think it proper for a source code explorer to dictate the "proper" licenses for Go package.

Of course, please disregard my comments if you consider free, manual legal review by Google an acceptable solution for every package that needs it. If so, I still think there should be an automatic, faster (and cheaper for Google?) way to smooth the process; hence my .allow-sourceexplorers strawman.

**Comment by julieqiu** [comment]
gioui.org will be fixed on our next deploy, so there isn't any further action needed on your part.

I will have to defer to our legal team on their decision for how to the AND vs OR decision for dual licenses because IANAL.

Our current volume of special case issues is low enough that we plan to keep the manual review process for now (rather than building an automatic solution). We can always revisit this strategy as other specific issues come up.

**Comment by mewmew** [comment]
> gioui.org will be fixed on our next deploy, so there isn't any further action needed on your part.

Good that gioui.org is now supported by pkg.go.dev. However, what about other packages that are dual-licensed, that are put in the public domain (through UNLICENSED) or otherwise blocked by pkg.go.dev?

There are currently [2,485 Go repositories released into the public domain](https://github.com/search?q=license%3Aunlicense+language%3AGo&type=Repositories) (through UNLICENSED). For comparison, the total number of open source Go repositories hosted on GitHub is 232,299, so ~1% are released into the public domain.

To give another comparison, there are more Go projects released into the public domain (through UNLICENSED) than there are [Go projects licensed under AGPL3](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aagpl-3.0&type=Repositories), of which there are currently 2,244. And AGPL3 is recognized by pkg.go.dev.

There are currently [39 Go repositories licensed as Artistic License 2.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aartistic-2.0&type=Repositories), another license recognized by pkg.go.dev.

For a full set of statistics, see below:

* `mit`: [MIT License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Amit&type=Repositories)
    - 112,974 (*approved* by `pkg.go.dev`)
* `apache-2.0`: [Apache License 2.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aapache-2.0&type=Repositories)
    - 45,046 (*approved* by `pkg.go.dev`)
* `gpl-3.0`: [GNU General Public License v3.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Agpl-3.0&type=Repositories)
    - 14,517 (*approved* by `pkg.go.dev`)
* `bsd-3-clause`: [BSD 3-Clause "New" or "Revised" License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Absd-3-clause&type=Repositories)
    - 8,157 (*approved* by `pkg.go.dev`)
* `bsd-2-clause`: [BSD 2-Clause "Simplified" License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Absd-2-clause&type=Repositories)
    - 4,059 (*approved* by `pkg.go.dev`)
* `gpl-2.0`: [GNU General Public License v2.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Agpl-2.0&type=Repositories)
    - 2,780 (*approved* by `pkg.go.dev`)
* `unlicense`: [The Unlicense](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aunlicense&type=Repositories)
    - 2,485 (**not** approved* by `pkg.go.dev`)
* `mpl-2.0`: [Mozilla Public License 2.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Ampl-2.0&type=Repositories)
    - 2,445 (*approved* by `pkg.go.dev`)
* `agpl-3.0`: [GNU Affero General Public License v3.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aagpl-3.0&type=Repositories)
    - 2,244 (*approved* by `pkg.go.dev`)
* `lgpl-3.0`: [GNU Lesser General Public License v3.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Algpl-3.0&type=Repositories)
    - 1,858 (*approved* by `pkg.go.dev`)
* `isc`: [ISC License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aisc&type=Repositories)
    - 1,433 (*approved* by `pkg.go.dev`)
* `wtfpl`: [Do What The F*ck You Want To Public License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Awtfpl&type=Repositories)
    - 552 (**not** approved* by `pkg.go.dev`)
* `cc0-1.0`: [Creative Commons Zero v1.0 Universal](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Acc0-1.0&type=Repositories)
    - 231 (**not** approved* by `pkg.go.dev`)
* `lgpl-2.1`: [GNU Lesser General Public License v2.1](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Algpl-2.1&type=Repositories)
    - 167 (*approved* by `pkg.go.dev`)
* `zlib`: [zlib License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Azlib&type=Repositories)
    - 162 (*approved* by `pkg.go.dev`)
* `epl-1.0`: [Eclipse Public License 1.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aepl-1.0&type=Repositories)
    - 110 (**not** approved* by `pkg.go.dev`)
* `epl-2.0`: [Eclipse Public License 2.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aepl-2.0&type=Repositories)
    - 46 (**not** approved* by `pkg.go.dev`)
* `bsl-1.0`: [Boost Software License 1.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Absl-1.0&type=Repositories)
    - 41 (*approved* by `pkg.go.dev`)
* `artistic-2.0`: [Artistic License 2.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aartistic-2.0&type=Repositories)
    - 39 (*approved* by `pkg.go.dev`)
* `cc-by-sa-4.0`: [Creative Commons Attribution Share Alike 4.0 International](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Acc-by-sa-4.0&type=Repositories)
    - 27 (**not** approved* by `pkg.go.dev`)
* `bsd-3-clause-clear`: [BSD 3-Clause Clear License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Absd-3-clause-clear&type=Repositories)
    - 18 (**not** approved* by `pkg.go.dev`)
* `osl-3.0`: [Open Software License 3.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aosl-3.0&type=Repositories)
    - 14 (**not** approved* by `pkg.go.dev`)
* `eupl-1.2`: [European Union Public License 1.2](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aeupl-1.2&type=Repositories)
    - 14 (**not** approved* by `pkg.go.dev`)
* `cc-by-4.0`: [Creative Commons Attribution 4.0 International](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Acc-by-4.0&type=Repositories)
    - 12 (**not** approved* by `pkg.go.dev`)
* `0bsd`: [BSD Zero Clause License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3A0bsd&type=Repositories)
    - 6 (**not** approved* by `pkg.go.dev`)
* `afl-3.0`: [Academic Free License v3.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aafl-3.0&type=Repositories)
    - 5 (**not** approved* by `pkg.go.dev`)
* `ms-pl`: [Microsoft Public License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Ams-pl&type=Repositories)
    - 3 (**not** approved* by `pkg.go.dev`)
* `eupl-1.1`: [European Union Public License 1.1](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aeupl-1.1&type=Repositories)
    - 2 (**not** approved* by `pkg.go.dev`)
* `bsd-4-clause`: [BSD 4-Clause "Original" or "Old" License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Absd-4-clause&type=Repositories)
    - 2 (**not** approved* by `pkg.go.dev`)
* `odbl-1.0`: [ODC Open Database License v1.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aodbl-1.0&type=Repositories)
    - 1 (**not** approved* by `pkg.go.dev`)
* `upl-1.0`: [Universal Permissive License v1.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aupl-1.0&type=Repositories)
    - 0 (**not** approved* by `pkg.go.dev`)
* `postgresql`: [PostgreSQL License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Apostgresql&type=Repositories)
    - 0 (**not** approved* by `pkg.go.dev`)
* `ofl-1.1`: [SIL Open Font License 1.1](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aofl-1.1&type=Repositories)
    - 0 (**not** approved* by `pkg.go.dev`)
* `ncsa`: [University of Illinois/NCSA Open Source License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Ancsa&type=Repositories)
    - 0 (**not** approved* by `pkg.go.dev`)
* `ms-rl`: [Microsoft Reciprocal License](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Ams-rl&type=Repositories)
    - 0 (**not** approved* by `pkg.go.dev`)
* `lppl-1.3c`: [LaTeX Project Public License v1.3c](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Alppl-1.3c&type=Repositories)
    - 0 (**not** approved* by `pkg.go.dev`)
* `ecl-2.0`: [Educational Community License v2.0](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Aecl-2.0&type=Repositories)
    - 0 (**not** approved* by `pkg.go.dev`)
* `cecill-2.1`: [CeCILL Free Software License Agreement v2.1](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3Acecill-2.1&type=Repositories)
    - 0 (**not** approved* by `pkg.go.dev`)

---

[Total number of open source Go repositories on GitHub](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo&type=Repositories):
- 293,872

Total number of open source Go repositories with license recognized GitHub search:
- 199,450

Total number of open source Go repositories with license approved by pkg.go.dev:
- 195,922

Total number of open source Go repositories with license not approved by pkg.go.dev:
- 3,528 (out of which 2,485 are projects in the public domain with UNLICENSED)

---

Personally, I think that this helped me gain a better understanding of the license landscape of the Go ecosystem. Prior to this quick evaluation, I felt that the list of licenses was too restrictive. However, given that it is possible to dual-license projects (although it will require manual effort to get them on pkg.go.dev), I think that the list of licenses is good enough as is.

I would very much like to see a public domain option added to the list, as it is *the* license used by most Go projects which is not yet supported by pkg.go.dev.

With kind regards,
Robin Eklind

**Comment by jba** [comment]
Thanks for that analysis. Very illuminating.

We appreciate what motivates people to reach for a public-domain style license, and in our own analysis, we've also identified that as a key issue. We'll try to have an update on that topic soon.

**Comment by networkimprov** [comment]
@mewmew do those numbers count github "forks"?

**Comment by mewmew** [comment]
> @mewmew do those numbers count github "forks"?

Good question. The number is directly taken from the linked GitHub searches. I did not think of further refining the search to account for forks (i.e. duplicates). Feel free to post a follow-up analysis :)

**Comment by mewmew** [comment]
> Edit: I do wonder though, what those 94,422 (32%) of the projects without a license recognized by GitHub search are.

I can give at least an approximate answer to this. At least part of the number includes dual-licensed projects. For instance, https://github.com/llir/llvm does not show up when searching for [0-clause BSD](https://github.com/search?utf8=%E2%9C%93&q=language%3AGo+license%3A0bsd&type=Repositories), but the project is dual-licensed into the public domain (through UNLICENSED) and under a 0-clause BSD license. The specific issue tracking this is #36983.

**Comment by logrusorgru** [comment]
> because it doesn't encourage package authors to fix their licenses for the broader open source ecosystem if there are issues

It's called "force", not "encourage". Please call things with their real names.

So, why I should be forced to use an "approved license"? I did my best to choose my license, having considered all the possible problems:

- for non English speakers too
- don't want to force users to keep copyright, license or care about it more then a second at life
- don't want to force users to read CC0 (near alternative) that's very-very-very long english text a regular human can't understand at all.

And I can conclude, that the limitation is just the **imposition** of Google/Go’s will.

**Comment by networkimprov** [comment]
@logrusorgru please file a new issue to request inclusion of your packages on go.dev.

**Comment by arp242** [comment]
> We (and the proxy team) considered that option and decided that it wasn't viable. We would rather not solve this problem with a "just for us" solution, because it doesn't encourage package authors to fix their licenses for the broader open source ecosystem if there are issues, and other users have concerns about a package or module's licenses.

I think the biggest problem here is that the LICENCE file is intended for humans to read and review, and not for programs to get data from.

Especially larger projects have more complex licensing than "just MIT" "just GPL"; they may be dual-licensed, have difference licenses for different files, and so forth. It's almost impossible to communicate this clearly in a LICENSE file *and* remain usable by pkg.go.dev.

I think "package authors should fix their license" is just *one* reason licenses aren't being detected, and to be honest I think the above is too dismissive of the complexities of real-world licensing @julieqiu. 

I don't know what the best solution for this is; I know there's https://reuse.software/spec, but it's fairly new without much adaptation (AFAIK), but it would solve the problem without a "just for us" solution and I think it's worth taking a look at. I don't know if there are other similar standards.

**Comment by logrusorgru** [comment]
> @logrusorgru please file a new issue to request inclusion of your packages on go.dev.

I have already got the reject by email some time ago.

> Me.: Hello. Please, add WTFPL license to the https://pkg.go.dev/license-policy list. For example https://github.com/logrusorgru/aurora you can see GitHub detects the...

> Jonathan Amsterdam <jba@google.com>: Thanks for your request, Konstantin. Unfortunately, we're not able to use WTFPL.

Regardless the github.com/google/licensecheck [detects it ok](https://github.com/google/licensecheck/tree/d5d259d76f74414dbc202b66ff1edf2a7de6fcb9/licenses).

**Comment by networkimprov** [comment]
Things may have changed since then. And you can't necessarily trust email :-)

**Comment by jba** [comment]
@logrusorgru, licensecheck's job is just to recognize licenses. pkg.go.dev has stronger requirements: the license must grant us permission to display the README and documentation, and must have the right legal properties. I'm afraid nothing has changed with regard to our treatment of WTFPL.

**Comment by jba** [comment]
> https://reuse.software/spec ... would solve the problem

@arp242, That spec mandates that a license file's name is the SPDX identifier for the license it contains. Presumably you're suggesting that we use the SPDX identifier and ignore the contents.

The problem is: what if they don't match? What if the license forbids redistribution but its filename says it's OK? It's unclear what the legal ramifications are, but even we non-lawyers can understand that a human judge might give more weight to a complete natural-language document than to a short name. 

And frankly, license recognition is not the biggest problem. We have a couple of ways to recognize non-standard license text.

**Comment by logrusorgru** [comment]
> @logrusorgru, licensecheck's job is just to recognize licenses. pkg.go.dev has stronger requirements: the license must grant us permission to display the README and documentation, and must have the right legal properties. I'm afraid nothing has changed with regard to our treatment of WTFPL.

So, I can find arguments against this position. But, seems, other corporations and users can stuck with the same fears. It's better to move my projects to Unlicensed, I think. It's the shortest and the simplest one of approved.

Thanks you @jba, @networkimprov.

**Comment by ddevault** [comment]
I am maintaining an independent fork of godoc.org at https://godocs.io, and will do so indefinitely. Feel free to use it if your project is being bullied into changing its license or having docs taken away from it.

**Comment by ddevault** [comment]
Sharing information about independent forks which address features that the Go team does not want to address, in issue threads about those same features, is not spam. Please provide more detail if you believe otherwise.

I get that you're proud of pkg.go.dev and uncomfortable with those who don't like it, but that doesn't change the facts.

**Comment by andybons** [comment]
Both comments are off topic from the discussion. You have other, more appropriate forums to advertise your service. Please use them. Thanks.

**Comment by ddevault** [comment]
Re-iterating is not the same thing as providing rationale. This is hardly an advertisement. The moderation tactics here smell of salty maintainers rather than dispassionate rules enforcement.
